cooltools/expected.py

```diff
@@ -987,6 +1377,287 @@ def blocksum_asymm(
+    DataFrame with summary statistic of every diagonal of every symmetric
+    or asymmetric block:
+    region1, region2, diag, n_valid, count.sum count.avg, etc
+
+    """
+
+    if view_df is None:
+        if not symmetric:
+            raise ValueError("asymmetric regions has to be smaller then full chromosomes, use view_df")
+        # Generate viewframe from clr.chromsizes:
+        view_df = bioframe.make_viewframe(
+            [(chrom, 0, clr.chromsizes[chrom]) for chrom in clr.chromnames]
+        )
+    else:
+        # Make sure view_df is a proper viewframe
+        try:
+            if not bioframe.is_viewframe(view_df, raise_errors=True):
+                raise ValueError("view_df is not a valid viewframe.")
+            if not bioframe.is_contained(view_df, bioframe.make_viewframe(clr.chromsizes)):
+                raise ValueError(
+                    "View table is out of the bounds of chromosomes in cooler."
+                )
+            # add is_sorted check in the asymmetric case ...
+        except Exception as e:  # AssertionError or ValueError, see https://github.com/gfudenberg/bioframe/blob/main/bioframe/core/checks.py#L177
+            warnings.warn(
+                "view_df has to be a proper viewframe from next release",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            view_df = bioframe.make_viewframe(view_df)
+
+    # define transforms - balanced and raw ('count') for now
+    if not balance:
+        # no transforms
+        clr_weight_name = None
+        transforms = {}
+    else:
+        if isinstance(balance, bool):
+            #weight default
+            clr_weight_name = "weight"
+        elif isinstance(balance, str):
+            #weight default
+            clr_weight_name = balance
+        else:
+            raise TypeError("balance has to be bool or str that specifies name of balancing weight in clr")
+        # define balanced data transform:
+        weight1 = clr_weight_name + "1"
+        weight2 = clr_weight_name + "2"
+        transforms = {"balanced": lambda p: p["count"] * p[weight1] * p[weight2]}
+
+    # check if clr_weight_name is in cooler
+    if clr_weight_name not in clr.bins().columns:
+        raise ValueError(f"specified balancing weight {clr_weight_name} is not available in cooler")
+
+    # execution details
+    if nproc > 1:
+        pool = mp.Pool(nproc)
+        map_ = pool.map
+    else:
+        map_ = map
+
+    # using try-clause to close mp.Pool properly
+    try:
+        if symmetric:
+            result = diagsum(
+                clr,
+                view_df,
+                transforms=transforms,
+                weight_name=clr_weight_name,
+                bad_bins=None,
+                chunksize=chunksize,
+                ignore_diags=ignore_diags,
+                map=map_,
+            )
+            # conform with the new expected format
+            # where each region is 2D, even if symmetric:
```

Comment:
actually hide the format transformation inside of `diagsum`. And consider making `region1/2` a categorical for lightness